### PR TITLE
Refactor #8792 to integrate latest master

### PR DESCRIPTION
**Pull request by jsuarez5341**
This PR adds batched compute_actions for MA inference and the FlexDict extra space for complex observations/actions. Details provided in #8792; this PR simply refactors to base off of the latest master

@sven1977 Could you take a look at this?

### CODE REVIEW

File: rllib/utils/spaces/space_utils.py [hunk 1]
```diff
@@ -21,10 +21,11 @@ def flatten_space(space):
     """
 
     def _helper_flatten(space_, l):
+        from ray.rllib.utils.spaces.flexdict import FlexDict
         if isinstance(space_, Tuple):
             for s in space_:
                 _helper_flatten(s, l)
-        elif isinstance(space_, Dict):
+        elif isinstance(space_, Dict) or isinstance(space_, FlexDict):
```
Comment (sven1977) [inline rllib/utils/spaces/space_utils.py:28]:
```suggestion
        elif isinstance(space_, (Dict, FlexDict)):
```

File: rllib/agents/trainer.py [hunk 2]
```diff
@@ -826,6 +828,97 @@ def compute_action(self,
         else:
             return result[0]  # backwards compatibility
 
+    def compute_actions(self,
```
Comment (sven1977) [inline rllib/agents/trainer.py:831]: Could we also deprecate `compute_action` and rename it into `compute_single_action`. Just add a `deprecation_warning("compute_action", "compute_single_action")` call at the beginning of `compute_action` and then return the results of `compute_single_action` from within there.

File: rllib/utils/spaces/flexdict.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+         self.assertSpace(space)
+
+      # None for shape and dtype, since it'll require special handling
+      self.np_random = None
+      self.shape     = None
+      self.dtype     = None
+      self.seed()
+
+   def assertSpace(self, space):
+      err = 'Values of the dict should be instances of gym.Space'
```
Comment (sven1977) [inline rllib/utils/spaces/flexdict.py:33]: This would make it impossible to have a `FlexDict` of `FlexDicts`, no?

**Review by sven1977 — CHANGES_REQUESTED** [review]

File: rllib/utils/spaces/space_utils.py (hunk 1 above)
Comment (jsuarez5341) [reply rllib/utils/spaces/space_utils.py:28]: Didn't know isinstance could take a tuple -- thanks!@

File: rllib/agents/trainer.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/trainer.py:831]: Actually, this is fine. We can do this quickly in a separate PR.

**Review by sven1977 — APPROVED** [review]
Thanks for the PR @jsuarez5341 .
Looks all good!

File: rllib/utils/spaces/flexdict.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        self.shape = None
+        self.dtype = None
+        self.seed()
+
+    def assertSpace(self, space):
+        err = "Values of the dict should be instances of gym.Space"
+        assert issubclass(type(space), gym.spaces.Space), err
+
+    def sample(self):
+        return dict([(k, space.sample()) for k, space in self.spaces.items()])
```
Comment (sven1977) [inline rllib/utils/spaces/flexdict.py:37]:
```suggestion
        return {k: space.sample() for k, space in self.spaces.items()}
```

Comment (jsuarez5341) [reply rllib/utils/spaces/flexdict.py]: Fixed

**Pull request merged**